tests/ndc/ndc.go

```diff
@@ -0,0 +1,2468 @@
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: workflowID,
+				RunId:      runID,
+			},
+		},
+	)
+	s.NoError(err)
+	if s.IsForceTerminated(workflowID, runID) {
+		index := resp.VersionHistories.CurrentVersionHistoryIndex
+		currentVersionHistory := resp.VersionHistories.Histories[index]
+		currentVersionHistory.Items[len(currentVersionHistory.Items)-1].EventId -= 1
+		// force termination event is generated by service itself, need to exclude
+	}
+	for _, actualVersionHistory := range resp.VersionHistories.Histories {
+		actualVersionHistory.BranchToken = nil
+	}
+	for _, actualVersionHistory := range resp.VersionHistories.Histories {
+		actualVersionHistory.BranchToken = nil
+		lcaItem, err := versionhistory.FindLCAVersionHistoryItem(
+			expectedVersionHistory,
+			actualVersionHistory,
+		)
+		s.NoError(err)
+		lastItem, err := versionhistory.GetLastVersionHistoryItem(expectedVersionHistory)
+		s.NoError(err)
+		if versionhistory.IsEqualVersionHistoryItem(lastItem, lcaItem) {
+			return
+		}
+	}
+	s.Fail(fmt.Sprintf(
+		"unable to find version history in mutable state %v vs %v",
+		expectedVersionHistory,
+		resp.VersionHistories.Histories,
+	))
+}
+
+func (s *NDCFunctionalTestSuite) verifyEventHistory(
+	workflowID string,
+	runID string,
+	historyBatch []*historypb.History,
+) {
+	// get replicated history events from passive side
+	replicatedHistory, err := s.cluster.GetFrontendClient().GetWorkflowExecutionHistory(
+		s.newContext(),
+		&workflowservice.GetWorkflowExecutionHistoryRequest{
+			Namespace: s.namespace.String(),
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: workflowID,
+				RunId:      runID,
+			},
+			MaximumPageSize:        1000,
+			NextPageToken:          nil,
+			WaitNewEvent:           false,
+			HistoryEventFilterType: enumspb.HISTORY_EVENT_FILTER_TYPE_ALL_EVENT,
+		},
+	)
+	s.NoError(err)
+
+	// compare origin events with replicated events
+	batchIndex := 0
+	batch := historyBatch[batchIndex].Events
+	eventIndex := 0
+	for _, event := range replicatedHistory.GetHistory().GetEvents() {
+		if eventIndex >= len(batch) {
+			batchIndex++
+			batch = historyBatch[batchIndex].Events
+			eventIndex = 0
+		}
+		originEvent := batch[eventIndex]
+		eventIndex++
+		s.Equal(originEvent.GetEventType(), event.GetEventType())
+	}
+}
+
+func (s *NDCFunctionalTestSuite) setupRemoteFrontendClients() {
+	s.mockAdminClient["cluster-b"].(*adminservicemock.MockAdminServiceClient).EXPECT().ReapplyEvents(gomock.Any(), gomock.Any()).Return(&adminservice.ReapplyEventsResponse{}, nil).AnyTimes()
+	s.mockAdminClient["cluster-c"].(*adminservicemock.MockAdminServiceClient).EXPECT().ReapplyEvents(gomock.Any(), gomock.Any()).Return(&adminservice.ReapplyEventsResponse{}, nil).AnyTimes()
+}
+
+func (s *NDCFunctionalTestSuite) sizeOfHistoryEvents(
+	events []*historypb.HistoryEvent,
+) int64 {
+	blob, err := serialization.NewSerializer().SerializeEvents(events, enumspb.ENCODING_TYPE_PROTO3)
+	s.NoError(err)
```

Comment:
If test calling sizeOfHistoryEvents fails, it will report error on this line.  Just return error from here and assert inside the test. This will also allow you have it as a function and not as a method.